lib/plugins/aws/package/compile/events/cognitoUserPool/index.js

```diff
@@ -0,0 +1,185 @@
+              // Check `cognitoUserPool` object has required properties
+              if (!event.cognitoUserPool.pool || !event.cognitoUserPool.trigger) {
+                throw new this.serverless.classes
+                  .Error([
+                    `Cognito User Pool event of function ${functionName} is not an object.`,
+                    'The correct syntax is an object with the "pool" and "trigger" properties.',
+                    'Please check the docs for more info.',
+                  ].join(' '));
+              }
+
+              // Check `cognitoUserPool` trigger is valid
+              if (!_.includes(validTriggerSources, event.cognitoUserPool.trigger)) {
+                throw new this.serverless.classes
+                  .Error([
+                    'Cognito User Pool trigger source is invalid, must be one of:',
+                    `${validTriggerSources.join(', ')}.`,
+                    'Please check the docs for more info.',
+                  ].join(' '));
+              }
+
+              // Save trigger functions so we can use them to generate
+              // IAM permissions later
+              cognitoUserPoolTriggerFunctions.push({
+                functionName,
+                poolName: event.cognitoUserPool.pool,
+                triggerSource: event.cognitoUserPool.trigger,
+              });
+
+              // Save user pools so we can use them to generate
+              // CloudFormation resources later
+              userPools.push(event.cognitoUserPool.pool);
+            } else {
+              throw new this.serverless.classes
+                .Error([
+                  `Cognito User Pool event of function ${functionName} is not an object.`,
+                  'The correct syntax is an object with the "pool" and "trigger" properties.',
+                  'Please check the docs for more info.',
+                ].join(' '));
+            }
+          }
+        });
+      }
+    });
+
+    // Generate CloudFormation templates for Cognito User Pool changes
+    _.forEach(userPools, (poolName) => {
+      // Create a `LambdaConfig` object for the CloudFormation template
+      const lambdaConfig = _.reduce(cognitoUserPoolTriggerFunctions, (result, value) => {
+        const lambdaLogicalId = this.provider.naming.getLambdaLogicalId(value.functionName);
+
+        // Return a new object to avoid lint errors
+        return Object.assign({}, result, {
+          [value.triggerSource]: {
+            'Fn::GetAtt': [
+              lambdaLogicalId,
+              'Arn',
+            ],
+          },
+        });
+      }, {});
+
+      const userPoolLogicalId = this.provider.naming.getCognitoUserPoolLogicalId(poolName);
+
+      const userPoolTemplate = {
+        Type: 'AWS::Cognito::UserPool',
+        Properties: {
+          UserPoolName: poolName,
+          LambdaConfig: lambdaConfig,
+        },
+      };
+
+      // Check if pool name has been declared in `Resources`
+      if (_.has(
```

Comment:
Should we maybe check for the `Type` as well? Might be pretty unlikely but I'm thinking about the case when one has defined another Resource (e.g. an S3 bucket) with the name of a pool.